### PR TITLE
feat: add app.getSystemLocale() method

### PR DESCRIPTION
**Pull request by rzhao271**
#### Description of Change

This PR adds a method to fetch the OS locale from Chromium. That way, even when the application is in a certain locale, the program can still know the system locale and make recommendations based off of that.

Downstream: https://github.com/microsoft/vscode/issues/159813 and PR https://github.com/microsoft/vscode/pull/159958.

CC @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added `app.getSystemLocale()` method.

### CODE REVIEW

File: spec/chromium-spec.ts [hunk 1]
```diff
@@ -394,6 +395,9 @@ describe('command line switches', () => {
 
       output = output.replace(/(\r\n|\n|\r)/gm, '');
       expect(output).to.equal(result);
+
+      const systemLocale = app.getSystemLocale();
```
Comment (rzhao271) [inline spec/chromium-spec.ts:399]: I'll see if I can move this so it's within `appProcess`.

File: shell/browser/electron_browser_main_parts.cc [hunk 2]
```diff
@@ -320,7 +320,11 @@ int ElectronBrowserMainParts::PreCreateThreads() {
   }
 #endif
 
-  // Initialize the app locale.
+  // Fetch the system locale for Electron.
+  std::string system_locale = l10n_util::GetApplicationLocale("", false);
+  fake_browser_process_->SetSystemLocale(system_locale);
```
Comment (ckerr) [inline shell/browser/electron_browser_main_parts.cc:325]:
```suggestion
  fake_browser_process_->SetSystemLocale(l10n_util::GetApplicationLocale("", false));
```

**Review by ckerr — APPROVED** [review]
Love it.

API LGTM

**Review by ckerr — APPROVED** [review]

File: docs/api/app.md [hunk 3]
```diff
@@ -723,6 +723,13 @@ Returns `string` - User operating system's locale two-letter [ISO 3166](https://
 
 **Note:** When unable to detect locale country code, it returns empty string.
 
+### `app.getSystemLocale()`
+
+Returns `string` - The current system locale, fetched using Chromium's `l10n_util` library.
+Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).
+
+**Note:** On Windows, you have to call it after the `ready` events gets emitted.
```
Comment (deepak1556) [inline docs/api/app.md:731]: Lets be consistent for all platforms and only allow calling this API after app ready. Also the API should throw an error if called before app ready to support this note. An example, https://github.com/electron/electron/blob/fa3cd1747556b66a3e714f99e988e47334187e44/shell/browser/api/electron_api_app.cc#L1189-L1193

File: shell/browser/electron_browser_main_parts.cc [hunk 4]
```diff
@@ -320,7 +320,11 @@ int ElectronBrowserMainParts::PreCreateThreads() {
   }
 #endif
 
-  // Initialize the app locale.
+  // Fetch the system locale for Electron.
+  fake_browser_process_->SetSystemLocale(
```
Comment (deepak1556) [inline shell/browser/electron_browser_main_parts.cc:324]: `l10n_util::GetApplicationLocale` will just return the preferred locale from the application bundle and not the actual system locale https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;l=491 . Also for other platforms the system locale is only obtained via ICU when the API could find an application preferred locale https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;l=528. I suggest the following changes instead,

1) Move `fake_browser_process_->SetSystemLocale` to the top of this function before `--lang` gets processed so that ICU default is not altered.

2) For windows and linux, call into `base::i18n::GetConfiguredLocale` which will obtain the value from `icu::Locale::getDefault` that does the necessary system calls ex: `GetLocaleInfoEx` refs, 
https://source.chromium.org/chromium/chromium/src/+/main:base/i18n/rtl.cc;l=96-99
https://source.chromium.org/chromium/chromium/src/+/main:third_party/icu/source/common/putil.cpp;l=1674-1951

3) For macOS, call into `[NSLocale currentLocale]`, refs https://developer.apple.com/documentation/foundation/nslocale/1409990-currentlocale . You might need to perform the `_` to `-` conversion still similar to https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util_mac.mm;l=41-45 but please test if that is needed.

**Review by deepak1556 — CHANGES_REQUESTED** [review]

File: shell/browser/electron_browser_main_parts.cc (hunk 4 above)
Comment (rzhao271) [reply shell/browser/electron_browser_main_parts.cc]: I have confirmed that the `_` to `-` conversion is still needed by writing a small test app that prints the locale in XCode (it printed `en_GB`).

File: docs/api/app.md (hunk 3 above)
Comment (deepak1556) [inline docs/api/app.md:731]:
```suggestion
**Note:** This API must be called after the `ready` event is emitted.
```

File: shell/browser/browser_process_impl.h [hunk 5]
```diff
@@ -100,7 +100,9 @@ class BrowserProcessImpl : public BrowserProcess {
   void StartAutoupdateTimer() override {}
 #endif
   void SetApplicationLocale(const std::string& locale) override;
+  void SetSystemLocale(const std::string& locale);
```
Comment (deepak1556) [inline shell/browser/browser_process_impl.h:103]: nit: move the non-virtual methods to the top before the overrides to be consistent.

**Review by deepak1556 — COMMENTED** [review]
LGTM pending API review and some minor style changes.

File: docs/api/app.md [hunk 6]
```diff
@@ -723,6 +723,13 @@ Returns `string` - User operating system's locale two-letter [ISO 3166](https://
 
 **Note:** When unable to detect locale country code, it returns empty string.
 
+### `app.getSystemLocale()`
+
+Returns `string` - The current system locale, fetched using Chromium's `l10n_util` library.
+Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).
```
[the remark this answers is not in the document]
Comment (itsananderson) [reply docs/api/app.md:729]: Also, looking at the docs for `app.getLocale`, it's not clear to me what it returns if you don't specify the `--lang` flag explicitly. Does it fall back to the system locale or just default to English? That probably would have been good to document before, but while we're adding `app.getSystemLocale` it would be good to clarify that default behavior of `app.getLocale`. If it returns the same value as `app.getSystemLocale`, it would probably be sufficient to note that fact and link to the `app.getSystemLocale` section.

File: shell/browser/electron_browser_main_parts.cc [hunk 7]
```diff
@@ -285,6 +286,13 @@ int ElectronBrowserMainParts::PreCreateThreads() {
   if (!views::LayoutProvider::Get())
     layout_provider_ = std::make_unique<views::LayoutProvider>();
 
+    // Fetch the system locale for Electron.
```
Comment (itsananderson) [inline shell/browser/electron_browser_main_parts.cc:289]: nit: is this indentation correct? Looks like it may be indented 2 more spaces than the `fake_browser_provess_` lines

File: docs/api/app.md [hunk 8]
```diff
@@ -715,14 +715,21 @@ To set the locale, you'll want to use a command line switch at app startup, whic
...
 Returns `string` - User operating system's locale two-letter [ISO 3166](https://www.iso.org/iso-3166-country-codes.html) country code. The value is taken from native OS APIs.
 
 **Note:** When unable to detect locale country code, it returns empty string.
 
+### `app.getSystemLocale()`
+
+Returns `string` - The current system locale, fetched using Chromium's `l10n_util` library.
+Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).
+
+**Note:** This API must be called after the `ready` event is emitted.
```
Comment (itsananderson) [inline docs/api/app.md:731]: Can you update the `app.getLocale` note to match this one?

Comment (rzhao271) [reply docs/api/app.md:731]: The `app.getLocale` API is older and doesn't have a check to ensure the caller is calling it after the `ready` event. The downside of adding the check is it might break platforms where one can get the correct result even before the `ready` event, but the upside is it could prevent [a few bug reports](https://github.com/electron/electron/issues?q=is%3Aissue+getLocale+is%3Aclosed+ready). If it is worth adding the check, I think that should go into a separate PR.

File: docs/api/app.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rzhao271) [reply docs/api/app.md:729]: For reference, `app.getLocale` calls into [l10n_util::GetApplicationLocale](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;drc=4c33ea7ce9a4064087f939723c0e9713193cf54f;l=573) which is much more involved.
All I know for now is that it might not return the same value as `app.getSystemLocale` by default. On my Windows machine, it returns the same value as `app.getSystemLocale`, but it seems like it can fall back to `en-US` on [GetApplicationLocaleInternalMac](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;l=498;drc=4c33ea7ce9a4064087f939723c0e9713193cf54f).

File: docs/api/app.md (hunk 8 above)
Comment (itsananderson) [reply docs/api/app.md:731]: I would still suggest updating the docs to state that the API must be called after the `ready` event, even if that isn't strictly enforced on macOS and Linux currently. I agree, though, that adding an enforcement check would be better to do in another PR.

File: docs/api/app.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ckerr) [reply docs/api/app.md:729]: :+1: on @itsananderson's suggestion of finding a version-friendly way of documenting the values for this new function and for `app.getLocale()` but it's out-of-scope for this PR

**Review by ckerr — COMMENTED** [review]
API LGTM.

File: shell/browser/electron_browser_main_parts.h [hunk 9]
```diff
@@ -130,6 +130,7 @@ class ElectronBrowserMainParts : public content::BrowserMainParts {
   void FreeAppDelegate();
   void RegisterURLHandler();
   void InitializeMainNib();
+  std::string GetCurrentSystemLocale();
```
Comment (ckerr) [inline shell/browser/electron_browser_main_parts.h:133]: I think this new method can be `static`?

**Review by ckerr — APPROVED** [review]
impl looks fine.

Deferring to @deepak1556's better wisdom on the macOS string underscores vs dashes. Looks like the new test code covers that, too.

**Comment by itsananderson** [comment]
API LGTM

File: shell/browser/electron_browser_main_parts_mac.mm [hunk 10]
```diff
@@ -74,4 +76,23 @@
...
+  std::string locale_value = [[systemLocaleIdentifier
+      stringByReplacingOccurrencesOfString:@"_"
+                                withString:@"-"] UTF8String];
+
+  // On disk the "en-US" resources are just "en" (http://crbug.com/25578), so
+  // the reverse mapping is done here to continue to feed Chrome the same values
+  // in all cases on all platforms.  (l10n_util maps en to en-US if it gets
+  // passed this on the command line)
+  if (locale_value == "en")
+    locale_value = "en-US";
```
Comment (deepak1556) [inline shell/browser/electron_browser_main_parts_mac.mm:93]:
```suggestion

```

This is only needed when obtaining the application locale because chrome ships `en.lproj` instead of `en_US.lpoj` https://bugs.chromium.org/p/chromium/issues/detail?id=25578

File: filenames.auto.gni [hunk 11]
```diff
@@ -140,6 +140,7 @@ auto_filenames = {
     "lib/common/define-properties.ts",
     "lib/common/ipc-messages.ts",
     "lib/common/web-view-methods.ts",
+    "lib/common/webpack-globals-provider.ts",
```
Comment (deepak1556) [inline filenames.auto.gni:143]: Seems unrelated ?

**Review by deepak1556 — APPROVED** [review]

**Review by MarshallOfSound — APPROVED** [review]
API LGTM
